enhancements/kubernetes-nmstate-integration.md

```diff
@@ -0,0 +1,165 @@
+These objects are created by kubernetes-nmstate and must not be touched by a user.
+`NodeNetworkConfigurationPolicy` objects can be used to specify desired
+networking state per node or set of nodes. It uses API similar to `NodeNetworkState`. 
+`NodeNetworkConfigurationEnactment` objects will be created per each node, per each matching
+`NodeNetworkConfigurationPolicy`.
+
+kubernetes-nmstate DaemonSet creates a custom resource of `NodeNetworkState` type per each node and
+updates the network topology from each OpenShift node.
+
+User configures host network changes and apply a policy in `NodeNetworkConfigurationPolicy` custom
+resource. Network topology is configured via `desiredState` section in `NodeNetworkConfigurationPolicy`.
+Multiple `NodeNetworkConfigurationPolicy` custom resources can be created.
+
+Upon receiving a notification event of `NodeNetworkState` update,
+kubernetes-nmstate Daemon verify the correctness of `NodeNetworkState` custom resource and
+apply the selected profile to the specific node.
+
+`NodeNetworkConfigurationEnactment` objects is a read-only object that represents Policy exectuion
+per each matching Node. It will expose configuration status per each Node.
+
+A new container image (kubernetes-nmstate-handler) will be created.
+
+A new container image (kubernetes-nmstate-operator) will be created. However, the operator
+and the handler will co-exist in the same upstream repo.
+
+An upstream API group of nmstate.io is currently used.
+
+### User Stories
+
+#### Bond creation
+
+* Be able to create bond interfaces on OpenShift nodes.
+ 
+  Typically, MCO + Ignition would be utilized. However, in a Day 1 install, that may not be possible.
+  It may be necessary for the bond/vlan configuration be done before Ignition can even be reached.
+  
+* Create a vlan interface on top of the bond inter.
+
+#### Assign ip address
+
+* Assign static and/or dynamic ip address on interfaces
+* Assign ipv4 and/or ipv6
+
+#### Create/Update/Remove network routes
+
+* Be able to Create/Update/Remove network routes for different interfaces like (bond,ethernet,sriov vf and sriov pf)
+
+#### Manage/Configure host SR-IOV network device
+
+* Be able to change host Virtual functions configuration (not managed by the sriov-operator) like vlans mtu driver etc..
+
+#### Rollback
+
+* Be able to rollback network configuration 
+if we lose connectivity to the openshift api server after applying a policy.
+
+### Implementation Details
+
+The proposal introduces kubernetes-nmstate as Tech Preview.
+
+## Design Details
+
+### Test Plan
+
+- Functional tests will be implemented
+
+### Graduation Criteria
+
+Initial support for kubernetes-nmstate will be Tech Preview
+
+#### Tech Preview
+
+- kubernetes-nmstate can be installed via container image
+- Host network topology can be configured via CRDs
+
+#### Tech Preview -> GA
+
+### Upgrade / Downgrade Strategy
+
+### Version Skew Strategy
```

Comment:
This needs to be filled out as well. How will the configuration change? What are the components affected here? What are the APIs between them? Are these versioned? This goes far beyond defining a CRD.

How tightly is this coupled to the version of NetworkManager? To the kernel? Will this work on RHEL7? Will users need to care? What if NetworkManager deprecates a configuration directive? Adds a new one?

How tightly is this coupled to the advanced CNI plugins? SR-IOV, OVN-Kubernetes, etc? Remember, component upgrades are not strictly ordered.

Reply:
There is a single strong dependency and that is between kubernetes-nmstate and the NetworkManager running on the host. The communication happens through D-Bus. NetworkManager API is backwards compatible, which makes kubernetes-nmstate forward compatible. When we release, we support certain version of NetworkManager and everything after it. nmstate release cycle (not kubernetes-nmstate) is tightly bound to RHEL and NetworkManager version, so the only thing we need to ensure on kubernetes-nmstate here, is that we don't push a version that would be newer than NetworkManager currently available on RHCOS.

There should be no coupling with kernel version apart from that done by NetworkManager.

This will not work on RHEL7 since RHEL7 has old versions of NetworkManager. Support for RHEL7 is not on nmstate roadmap.

Users don't need to care. We just need to make sure to release only nmstate that is compatible with RHCOS/RHEL supported by the current version of OpenShift (with the exception of RHEL7). Our operator will not deploy kubernetes-nmstate on RHEL7 nodes.

In case there is a breaking change on NetworkManager side, we can tackle it in nmstate or kubernetes-nmstate with a workaround. However, since there is guaranteed backward compatibility, it would be a NetworkManager bug.

It is not coupled with any CNI. kubernetes-nmstate is just controlling configuration making sure to reach desired state. It won't intervene into any CNI configuration unless it is asked to.